races_test.go

```diff
@@ -0,0 +1,48 @@
+// +build race
+
+package eventhub
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+)
+
+func (suite *eventHubSuite) TestConcurrency() {
+	tests := map[string]func(context.Context, *testing.T, *Hub, string){
+		"TestConcurrentSendWithRecover": testConcurrentSendWithRecover,
+	}
+
+	for name, testFunc := range tests {
+		setupTestTeardown := func(t *testing.T) {
+			hub, cleanup := suite.RandomHub()
+			defer cleanup()
+			partitionID := (*hub.PartitionIds)[0]
+			client, closer := suite.newClient(t, *hub.Name, HubWithPartitionedSender(partitionID))
+			defer closer()
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+			defer cancel()
+			testFunc(ctx, t, client, partitionID)
+		}
+
+		suite.T().Run(name, setupTestTeardown)
+	}
+}
+
+func testConcurrentSendWithRecover(ctx context.Context, t *testing.T, client *Hub, _ string) {
+	var wg sync.WaitGroup
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			// we don't check for errors here as the call to Recover()
+			// can cancel any in-flight calls to Send().  this is only
+			// an interesting test when race detection is enabled.
+			client.Send(ctx, NewEventFromString("Hello!"))
+			client.sender.Recover(ctx)
```

Comment:
With that said, it's probably still useful to collect any errors from .Recover(), right? Could we use a channel perhaps? I think our expectation is that all the Recover() calls work and do not fail, even if they might do some slightly redundant work.

Reply:
Hmm yes good point, fixed.